pkg/services/datasources/sql.go

```diff
@@ -0,0 +1,211 @@
+package datasources
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+
+	"github.com/go-xorm/xorm"
+
+	"github.com/grafana/grafana/pkg/components/securejsondata"
+	"github.com/grafana/grafana/pkg/infra/metrics"
+	"github.com/grafana/grafana/pkg/models"
+)
+
+type datasourceSqlImpl struct {
+	sqlStore *sqlstore.SqlStore
+	ctx      context.Context
+}
+
+func (ds *datasourceSqlImpl) GetDataSourceById(id int64, orgId int64) (*models.DataSource, error) {
+	var datasource *models.DataSource
+	err := ds.sqlStore.WithTransactionalDbSession(ds.ctx, func(sess *sqlstore.DBSession) error {
```

Comment:
Every sql should be wrapped in transaction that reuses any existing transaction from given context.